src/wallet/rpcwallet.cpp

```diff
@@ -1957,9 +1959,11 @@ static UniValue walletpassphrase(const JSONRPCRequest& request)
     // wallet before the following callback is called. If a valid shared pointer
     // is acquired in the callback then the wallet is still loaded.
     std::weak_ptr<CWallet> weak_wallet = wallet;
-    pwallet->chain().rpcRunLater(strprintf("lockwallet(%s)", pwallet->GetName()), [weak_wallet] {
+    pwallet->chain().rpcRunLater(strprintf("lockwallet(%s)", pwallet->GetName()), [weak_wallet, relock_time] {
         if (auto shared_wallet = weak_wallet.lock()) {
             LOCK(shared_wallet->cs_wallet);
+            // Skip if this is not the most recent rpcRunLater callback.
+            if (shared_wallet->nRelockTime != relock_time) return;
```

Comment:
I think this could introduce a bug where wallet never relocks.

In case where walletpassphrase is called from two threads, first thread sets `pwallet->nRelockTime` and is interrupted after constructing lambda but before invoking rpcRunLater. Second thread sets `pwallet->nRelockTime` to a different value and calls `rpcRunLater`, and then first thread wakes up and calls `rpcRunLater`.

Lambda from second thread never runs because lambda from first thread replaced it. But lambda from first thread has the wrong `relock_time` value so never relocks the wallet.

Simplest fix might be adding relock time to timer names "lockwallet(%s,%i)" above to make them unique

Reply:
That would make timers list grow - they are never removed, just replaced.

Reply:
Added a mutex to prevent concurrent calls with the same wallet. The lock is held during wallet `Unlock` and RPCRunLater` so the case you describe can't happen.

Reply:
Nice catch BTW.